lib/zzhc/zzhc_port.c

```diff
@@ -0,0 +1,365 @@
+		size_t slen)
+{
+	int i = 0, rc;
+
+	if (dlen == 0 || slen == 0) {
+		return 0;
+	}
+
+	rc = base64_encode(dst, dlen, &i, src, slen);
+	return (rc == 0) ? i : 0;
+}
+
+
+static int settings_set(const char *key, size_t len,
+		settings_read_cb read_cb, void *cb_arg)
+{
+	int rc;
+
+	if (strcmp(key, "iccid") == 0) {
+		rc = read_cb(cb_arg, ueiccid, ICCID_LEN);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/**@brief Initialization of external modules */
+static int ext_init(struct zzhc *ctx)
+{
+	int rc;
+	struct settings_handler *this_settings;
+
+	memset(ueiccid, 0, ICCID_LEN);
+
+	/* Allocate memory for thread, sem, at_list and storage */
+	ctx->thread = this_malloc(sizeof(struct k_thread));
+	if (ctx->thread == NULL) {
+		return -ENOBUFS;
+	}
+
+	ctx->sem = this_malloc(sizeof(struct k_sem));
+	if (ctx->sem == NULL) {
+		return -ENOBUFS;
+	}
+
+	ctx->at_list = this_malloc(sizeof(struct at_param_list));
+	if (ctx->at_list == NULL) {
+		return -ENOBUFS;
+	}
+
+	ctx->storage = this_malloc(sizeof(struct settings_handler));
+	if (ctx->storage == NULL) {
+		return -ENOBUFS;
+	}
+
+	/* Initialize AT-command list, which is used in parser */
+	rc = at_params_list_init(ctx->at_list, AT_PARAMS_MAX);
+	if (rc != 0) {
+		LOG_DBG("at_params_list_init() = %d", rc);
+		return rc;
+	}
+
+	/* Initialize settings. */
+	rc = settings_subsys_init();
+	if (rc) {
+		LOG_DBG("settings_subsys_init() = %d", rc);
+		return rc;
+	}
+
+	memset(ctx->storage, 0, sizeof(struct settings_handler));
+	this_settings = (struct settings_handler *)ctx->storage;
+	this_settings->name  = (char *)settings_name;
+	this_settings->h_set = settings_set;
+
+	rc = settings_register(this_settings);
+	if (rc) {
+		LOG_DBG("settings_register() = %d", rc);
+		return rc;
+	}
+
+	settings_load();
+
+	return 0;
+}
+
+/**@brief Un-initialization of external modules */
+static void ext_uninit(struct zzhc *ctx)
+{
+	this_free(ctx->sem);
+	at_params_list_free(ctx->at_list);
+	this_free(ctx->at_list);
+}
+
+/**@brief Decode json object.
+ *
+ * For more information on cJSON library, please visit:
+ * https://github.com/DaveGamble/cJSON
+ */
+static cJSON *json_object_decode(cJSON *obj, const char *str)
+{
+	return obj ? cJSON_GetObjectItem(obj, str) : NULL;
+}
+
+/**@brief Check HTTP payload.
+ *
+ * This function parses the return string from zzhc server as json format. If
+ * "resultCode" is "0" and "resultDesc" is "Success", then the returned result
+ * is regarded as success. In case where server returns something else, then
+ * the returned result is regarded as failure.
+ *
+ * @retval true  If "resultCode" is "0" and "resultDesc" is "Success".
+ * @retval false Else.
+ *
+ * For more information on cJSON library, please visit:
+ * https://github.com/DaveGamble/cJSON
+ */
+static bool check_http_payload(struct zzhc *ctx)
+{
+	bool resultCode = false, resultDesc = false;
+
+	cJSON *root_obj;
+	cJSON *resultCode_obj;
+	cJSON *resultDesc_obj;
+
+	cJSON_Init();
```

Comment:
Would it be possible to use Zephyr json library instead? https://github.com/zephyrproject-rtos/zephyr/blob/master/include/data/json.h

The cJSON library was brought into NCS as a dependency for nRF Cloud library, and the long-term plan is to drop its usage in favour of Zephyr json lib.

Reply:
Good idea. (since there is example to learn in NCS)
I'll try to re-implement check_http_payload() using Zephyr json library in the next push.

Reply:
FYI `cJSON` will be used by `aws` libraries. The reason this is done instead of zephyr json is that the zephyr json does not support floating point (I have been told), and floats are required by the thingy project (which makes sense). Maybe we should take this discussion elsewhere, but thought you should know.

Reply:
Ok, I was not aware, so it could be that my comment is not a valid one.

Still, in my opinion, it makes little sense to maintain an external library, if we have a native counterpart in Zephyr. If it is missing a feature, it could be extended.

Reply:
In the last push on 4th Nov 2019, cJSON has been replaced by zephyr's json library.